Allow user-specified underlying type for enums

IISResetMe opened the pull request:
## PR Summary
Fix #8028 

This change adds support for specifiying the underlying type for an enum:

```powershell
enum MyEnum : long 
{
  A = 0x0FFFFFFFFFFFFFFF
  B
}
# or
enum MyByte : byte 
{
  A = 0x01
  B = 0x02
  C = 0x03
  D
}
```

TODO:
 - [x] re-introduce range validation (fixed in f417453b99636bdb1fd10b25d3ea554304d72315)
   - [x] ~~currently breaks on negative value literals~~ (fixed in 19b277054a16139678b64f277e5dd95530638add)
 - [x] better parsing errors for invalid underlying types (fixed in fc5a608f3483941f82d1a817cbde1eead3c18a77)
 - [x] ~~value assignment is currently broken from non-int types~~ (fixed in f58eba81d585a74cdef8e22807e9f39710e406ea)
 - [x] negative parsing tests

<!-- summarize your PR between here and the checklist -->

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

IISResetMe committed:
Add basic enum basetype tests

IISResetMe committed:
Add optional base type to enum parsing rule

IISResetMe committed:
WIP: Variable underlying types for enum definitions

This commit changes the behavior of the DefineEnumHelper, specifically the way the underlying type of the Enum type is specified.
06ff8a3d makes the parser recognize a base type token, this changes the underlying type of the generated EnumBuilder if the base type token corresponds to a builtin CLS value type as specified in ECMA-335 §I.8.5.2 (CLS Rule 7), otherwise it defaults to [int]

TODO:
 - re-introduce range validation
 - better parsing errors for invalid underlying types

IISResetMe committed:
Add support for Flags() and fix implied member literal values

IISResetMe committed:
Fix CodeFactor issue (SA1116)

IISResetMe committed:
Add type-aware range check for enum member values

I'm thinking there must be an easier way of abstracting this away, but I can't think of anything

IISResetMe committed:
Reset tokenizer mode in EnumDefinitionRule

Forgot to reset the tokenizer in a previous commit, causing parsing errors.

IISResetMe committed:
Update error strings, add null check for Enum member values

IISResetMe committed:
Fix CodeFactor issue (SA1000, SA1019)

IISResetMe committed:
Fix valueTooBig test (off-by-one) in DefineEnum()

IISResetMe committed:
Fix type names in enum test

IISResetMe committed:
Add more enum tests

Added new context for testing behavior when Flags() is applied.
Increased test coverage for all underlying types, added type constraint test

IISResetMe committed:
Better error handling + parser errors for invalid enum underlying type

enums can only be based off builtin integral types, parser should report an error if any other kind of type name is passed in

iSazonov commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -1091,10 +1092,29 @@ internal static List<DefineEnumHelper> Sort(List<DefineEnumHelper> defineEnumHel
+                foreach (var type in _enumDefinitionAst.BaseTypes)
+                {
+                    var resolvedType = type.TypeName.GetReflectionType();
+                    if (resolvedType == null || !_validBaseTypes.Contains(resolvedType))
```
I think we could do the check in parser. Seems it is more useful.

iSazonov replied:
Our goal is to move as many checks to the parser as possible. These checks are no longer needed in PSTypes.

iSazonov commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -1150,14 +1168,29 @@ internal void DefineEnum()
-                        if (value < int.MaxValue)
+                    }
+
+                    if (_enumDefinitionAst.Attributes.Any(attr => attr.TypeName.GetReflectionType() == typeof(FlagsAttribute)) && value != 0)
```
```suggestion
                    if (value != 0 && _enumDefinitionAst.Attributes.Any(attr => attr.TypeName.GetReflectionType() == typeof(FlagsAttribute)))
```

iSazonov commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -1167,6 +1200,42 @@ internal void DefineEnum()
+
+            private void GetMaxValue(Type underlyingType, out ulong maxValue)
+            {
+                if (underlyingType == typeof(byte))
```
Please use switch statement.

iSazonov commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -1091,10 +1092,29 @@ internal static List<DefineEnumHelper> Sort(List<DefineEnumHelper> defineEnumHel
+                    if (resolvedType == null || !_validBaseTypes.Contains(resolvedType))
+                    {
+                        _parser.ReportError(type.Extent,
+                            nameof(ParserStrings.IntegralTypeExpected),
```
Please align arguments to one column (under `type.Extent` here).
Below too.

iSazonov commented on src/System.Management.Automation/resources/ParserStrings.resx:
```diff
@@ -1316,6 +1316,9 @@ ModuleVersion : Version of module to import. If used, ModuleName must represent
     <value>Type name expected.</value>
   </data>
+  <data name="IntegralTypeExpected" xml:space="preserve">
+    <value>'{0}' is not a valid integral type (one of byte, sbyte, short, ushort, int, uint, long or ulong)</value>
```
Maybe replace `integral type` with `underlying type`?

iSazonov commented on test/powershell/Language/Classes/scripting.enums.tests.ps1:
```diff
@@ -87,4 +119,5 @@ Describe 'Basic enum errors' -Tags "CI" {
     ShouldBeParseError 'enum foo { e = $foo }' EnumeratorValueMustBeConstant 15 -SkipAndCheckRuntimeError
     ShouldBeParseError 'enum foo { e = "hello" }' CannotConvertValue 15 -SkipAndCheckRuntimeError
     ShouldBeParseError 'enum foo { a;b;c;' MissingEndCurlyBrace 10
+    ShouldBeParseError 'enum foo : string { a }' IntegralTypeExpected 11 -SkipAndCheckRuntimeError
```
The same.

iSazonov committed:
Optimize counter branch when defining enum member values

Co-Authored-By: IISResetMe <IISResetMe@users.noreply.github.com>

IISResetMe committed:
Add enum type constraint check to parser

Given that valid enum underlying types are all resolvable at parse-time (and unlikely to change), we might as well have the Parser error when an invalid underlying type is passed in as a type constraint.

IISResetMe committed:
Merge branch 'feature/enum-with-underlying-type' of https://github.com/IISResetMe/PowerShell into feature/enum-with-underlying-type

IISResetMe committed:
Update enum type constraint test

ErrorId updated in @548b6cd3

iSazonov commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4532,6 +4532,8 @@ private StatementAst EnumDefinitionRule(List<AttributeBaseAst> customAttributes,
             //G      enum-member  new-lines:opt
             //G      enum-member-list   enum-member
 
+            Type[] validBaseTypes = new[] { typeof(byte), typeof(sbyte), typeof(short), typeof(ushort), typeof(int), typeof(uint), typeof(long), typeof(ulong) };
```
Why not static?
And if we moved this to Parser why we did not remove this from PSType?

iSazonov replied:
I think it should be static and we should use consts like TypeCode.Int16 and method type.GetTypeCode().

IISResetMe replied:
Switching to TypeCode constants sounds like an excellent idea, will try this

IISResetMe committed:
Remove redundant enum type check from PSType

IISResetMe committed:
Clean up DefineEnum()

Since (any valid) input value and maxValue are bound to be of the same type, we may as well mark maxValue dynamic as well - they will be comparable at runtime, and this avoids incorrectly erring when the user specifies a negative integer value for a member

IISResetMe committed:
Fix CodeFactor issues (SA1116, SA1001)

vexx32 commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -1091,11 +1091,16 @@ internal static List<DefineEnumHelper> Sort(List<DefineEnumHelper> defineEnumHel
-                int value = 0;
+
+                dynamic value = 0;
+                dynamic maxValue = underlyingType.GetField("MaxValue", BindingFlags.Public | BindingFlags.Static | BindingFlags.GetField).GetValue(null);
```
Given that there is a finite number of acceptable types that are permitted here, wouldn't using a type-based switch here outperform reflection for this lookup?

IISResetMe replied:
I don't think performance is a major concern here - it's retrieved _once_ per `enum` defined.

vexx32 replied:
Every Little Bit Counts™️  😄 

(But yeah, I guess that makes sense!)

iSazonov replied:
I'm sure the _engine_ should get rid of slow patterns even if they are supposedly rarely used.
We could use the pattern (type.GetTypeCode() and TypeCode.Int16):
https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs#L43-L57

iSazonov commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -1104,28 +1109,22 @@ internal void DefineEnum()
                         if (IsConstantValueVisitor.IsConstant(enumerator.InitialValue, out constValue, false, false))
                         {
-                            if (constValue is int)
+                            if (!LanguagePrimitives.TryConvertTo(constValue, underlyingType, out value))
```
Seems we could move the checks to parser too. Perhaps not in the PR.

iSazonov commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4543,53 +4545,94 @@ private StatementAst EnumDefinitionRule(List<AttributeBaseAst> customAttributes,
+                    {
+                        ReportIncompleteInput(After(colonToken),
+                            nameof(ParserStrings.TypeNameExpected),
+                            ParserStrings.TypeNameExpected);
```
Please align all arguments to one column.
Below too (there is many cases).

IISResetMe replied:
Fixed all instances of `ReportError()` and `ReportIncompleteInput()`

iSazonov commented on src/System.Management.Automation/resources/ParserStrings.resx:
```diff
@@ -1316,6 +1316,9 @@ ModuleVersion : Version of module to import. If used, ModuleName must represent
     <value>Type name expected.</value>
   </data>
+  <data name="InvalidUnderlyingType" xml:space="preserve">
+    <value>'{0}' is not a valid underlying type for enums. Expected a builtin integral type (one of byte, sbyte, short, ushort, int, uint, long or ulong)</value>
```
integral -> underlying

IISResetMe replied:
Name of the error string is already updated, and the error text clearly states that it's because of an underlying type constraint

iSazonov replied:
There is 'Expected a builtin __integral__ type'.

IISResetMe replied:
No, that's quite intentional. "builtin underlying type" would be ambiguous.

IISResetMe committed:
Add enum negative value tests

IISResetMe committed:
Refactor enum underlying type check

IISResetMe committed:
Change enum type parser check to TypeCode

IISResetMe committed:
Fix ReportError()/ReportIncompleteInput() argument alignment in Parser

iSazonov commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -1091,14 +1091,50 @@ internal static List<DefineEnumHelper> Sort(List<DefineEnumHelper> defineEnumHel
+                        maxValue = ulong.MaxValue;
+                        break;
+                    default:
+                        _parser.ReportError(
```
The same request - could we move the type check to parser?

IISResetMe replied:
I honestly don't think that's a good idea. Injecting dynamic properties (we don't know the underlying type at compile time) into the AST - the AST being publicly exposed - sounds like asking for trouble. And we're emitting parser errors related to type constraints in the TypeDefiner for classes, it's not like this behavior is inconsistent with the existing implementation

iSazonov replied:
Enum looks like const. So we can make all checks at parse time.

IISResetMe replied:
I know that we _can_, I'm saying we _shouldn't_ - if I apply [@lzybkr's line of reasoning here](https://github.com/IISResetMe/PowerShell/commit/bfd4be766246bdf33a0ffbb027ba66723fe06b49#r31450712), we'll at the very least have to keep the check in place here even if we also check at parse-time - and I'm not sure how useful that is

iSazonov replied:
Thanks for clarify! I see the point and agree that the design is good.

iSazonov commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4460,7 +4576,8 @@ private MemberAst ClassMemberRule(string className, out List<Ast> astsOnError)
                 // We have the start of a member, but didn't see a variable or 'def'.
-                ReportIncompleteInput(After(ExtentFromFirstOf(lastAttribute)),
+                ReportIncompleteInput(
+                    After(ExtentFromFirstOf(lastAttribute)),
```
@IISResetMe Please move all formatting changes in separate PR. Our policy is avoid mixing style and functional changes.

IISResetMe replied:
Ah, misunderstood your previous comment then

iSazonov replied:
My request was only for your new or changed code but if you fixed all code please move this to new PR and keep current PR as small as possible.

IISResetMe replied:
Done, will fix the indentation style issues in a separate PR

IISResetMe committed:
Revert "Fix ReportError()/ReportIncompleteInput() argument alignment in Parser"

This reverts commit 434bbab6bef1862ea125aad97f3eeeba37000657.
Will re-add in a separate PR

IISResetMe committed:
Simplify type names

vexx32 commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -1150,14 +1189,29 @@ internal void DefineEnum()
+
+                    if (value != 0 && _enumDefinitionAst.Attributes.Any(attr => attr.TypeName.GetReflectionType() == typeof(FlagsAttribute)))
+                    {
+                        if (value < maxValue / 2)
```
Shouldn't this be value <= maxValue / 2? 🙂 (or is there a catch I'm unaware of where you can't have an enum value that's equal to the MaxValue of the underlying type?)

IISResetMe replied:
You're right, it should. Just turns out that that doesn't matter here, because the highest cardinal number you can fit into any integral type is never even :)

```powershell
@([byte],[sbyte],[uint16],[int16],[int],[uint32],[long],[uint64]).ForEach({$_::MaxValue})
255
127
65535
32767
2147483647
4294967295
9223372036854775807
18446744073709551615
```

iSazonov commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4532,6 +4532,8 @@ private StatementAst EnumDefinitionRule(List<AttributeBaseAst> customAttributes,
             //G      enum-member  new-lines:opt
             //G      enum-member-list   enum-member
 
+            TypeCode validUnderlyingTypeCodes = TypeCode.Byte | TypeCode.Int16 | TypeCode.Int32 | TypeCode.Int64 | TypeCode.SByte | TypeCode.UInt16 | TypeCode.UInt32 | TypeCode.UInt64;
```
It should be `readonly static`.

lzybkr replied:
`const` would be preferred if it's allowed.

IISResetMe replied:
@iSazonov the type constraint for enums is defined as part of the Common Language Specification in ECMA-335 and hasn't changed since 2001 - I think it's unlikely we'll ever have to update it

lzybkr replied:
Moreover - it's a public api, the values can't change. This is why we also specify enum values in PowerShell in places where it might not seem necessary, e.g. `TokenKind`.

iSazonov commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4543,53 +4545,94 @@ private StatementAst EnumDefinitionRule(List<AttributeBaseAst> customAttributes,
+                        if (resolvedType == null || !validUnderlyingTypeCodes.HasFlag(resolvedType.GetTypeCode()))
+                        {
+                            ReportError(underlyingType.Extent,
+                                nameof(ParserStrings.InvalidUnderlyingType),
```
Please align arguments in the new code.
(Please push your removed style commit in new PR)

IISResetMe replied:
Done. Will submit new PR once this is done (to avoid merge conflicts with modified code paths in `Parser.cs`)

iSazonov commented on test/powershell/Language/Classes/scripting.enums.tests.ps1:
```diff
@@ -70,6 +70,50 @@ Describe 'enums' -Tags "CI" {
+        enum V1 {
+            A = -4
+            B
+        }
```
From line 41. This should works:
```powershell
enum V1 {
    A = -4
    B = [int]::MinValue
    C = -1
    D
}
```

IISResetMe replied:
Difference is that we're counting _towards_ zero when the previous value is negative, so the test on line 41 doesn't make much sense in relation to `Int.MinValue`. 

That being said, we should accept `Int.MinValue`, so I added a test for that + next consecutive auto-assignment

lzybkr commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -1091,11 +1091,52 @@ internal static List<DefineEnumHelper> Sort(List<DefineEnumHelper> defineEnumHel
+
+                dynamic value = 0;
+                dynamic maxValue = 0;
+                switch (Type.GetTypeCode(underlyingType))
```
Uh, it used to do something different for enums I think, you'll have to use your favorite decompiler on an older version of PowerShell to double check though.

IISResetMe replied:
In this case that isn't so interesting because we're always passing in a `Type` instance (for which every other instance I can find `Type.GetTypeCode()` is also used)

iSazonov replied:
I thought it is for future enhancement but if it was removed we can ignore this.

@IISResetMe Please don't set "Resolve conversation" until we finish the code review - it is difficult to understand that comments needs an attention. Please leave this to comment's author or maintainers.

IISResetMe replied:
@iSazonov got it, sorry about that

IISResetMe committed:
Update EnumDefinitionRule syntax description

Updated syntax description to reflect changes made.
Additionally marked `validUnderlyingTypeCodes` const.

IISResetMe committed:
Fix argument indentation

Aligned arguments passed to ReportError() and ReportIncompleteInput() in a single column in the code paths modified

IISResetMe committed:
Expand test for enum with negative initial value

IISResetMe committed:
Remove binary enum value incrementations when Flags() is present

vexx32 commented:
@IISResetMe for future reference when you look at adding that `[Flags()]` stuff back in, I think you could probably safely use bit shifts rather than `*= 2` 😉

IISResetMe committed:
Remove test for binary enum value incrementations when Flags() is present

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this and is fine with the syntax

IISResetMe commented on test/powershell/Language/Classes/scripting.enums.tests.ps1:
```diff
@@ -87,4 +121,5 @@ Describe 'Basic enum errors' -Tags "CI" {
     ShouldBeParseError 'enum foo { e = $foo }' EnumeratorValueMustBeConstant 15 -SkipAndCheckRuntimeError
     ShouldBeParseError 'enum foo { e = "hello" }' CannotConvertValue 15 -SkipAndCheckRuntimeError
     ShouldBeParseError 'enum foo { a;b;c;' MissingEndCurlyBrace 10
+    ShouldBeParseError 'enum foo : string { a }' InvalidUnderlyingType 11 -SkipAndCheckRuntimeError
```
@TravisEz13 Do we want to update `ParserStrings.EnumeratorValueTooLarge` is "either too large _or_ too small for a [typename]". 

Currently (since 5.0) we throw:

```powershell
PS C:\>enum E { A = [long]::MinValue }
At line:1 char:13
+ enum E { A = [long]::MinValue }
+              ~~~~~~~~~~~~~~~~
Enumerator value is too large for a System.Int.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : EnumeratorValueTooLarge
```

adityapatwardhan commented:
@TravisEz13 Please update your review.

vexx32 commented:
@TravisEz13 @IISResetMe apart from the merge conflict, was there anything else pending here? 💖

IISResetMe committed:
Rename ParserStrings.EnumeratorValueTooLarge

The contexts in which this error string is used covers not only literals that are too large, but also sometimes too small
Renamed to EnumeratorValueOutOfBounds

IISResetMe committed:
Add test for enum literal value too small

iSazonov commented:
@IISResetMe Please resolve merge conflicts.

iSazonov commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4524,72 +4524,122 @@ private StatementAst EnumDefinitionRule(List<AttributeBaseAst> customAttributes,
             //G      enum-member  new-lines:opt
             //G      enum-member-list   enum-member
 
+            const TypeCode validUnderlyingTypeCodes = TypeCode.Byte | TypeCode.Int16 | TypeCode.Int32 | TypeCode.Int64 | TypeCode.SByte | TypeCode.UInt16 | TypeCode.UInt32 | TypeCode.UInt64;
```
Const should be upper case:
```suggestion
            const TypeCode ValidUnderlyingTypeCodes = TypeCode.Byte | TypeCode.Int16 | TypeCode.Int32 | TypeCode.Int64 | TypeCode.SByte | TypeCode.UInt16 | TypeCode.UInt32 | TypeCode.UInt64;
```

iSazonov commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4524,72 +4524,122 @@ private StatementAst EnumDefinitionRule(List<AttributeBaseAst> customAttributes,
+                var enumDefn = new TypeDefinitionAst(extent, name.Value, customAttributes == null ? null : customAttributes.OfType<AttributeAst>(), members, TypeAttributes.Enum, underlyingTypeConstraint == null ? null : new[] { underlyingTypeConstraint });
+                if (customAttributes != null && customAttributes.OfType<TypeConstraintAst>().Any())
+                {
+                    //no need to report error since there is error reported in method StatementRule
```
Please add space:
```suggestion
                    // No need to report error since there is error reported in method StatementRule.
```

IISResetMe committed:
Fix comment + variable casing

IISResetMe committed:
Merge branch 'master' into feature/enum-with-underlying-type

IISResetMe committed:
Fix variable casing for ValidUnderlyingTypeCodes

IISResetMe commented:
@iSazonov done!

IISResetMe commented:
[The Codacy issue currently failing](https://app.codacy.com/app/slee/PowerShell/file/28451399091/issues/source?bid=10638692&fileBranchId=10639194#l1172) is not really relevant.

While the compiler can't prove it, that path is unreachable since `case TypeCode.Int32` will always be chosen if `typeConstraintAst == null`
